### ISSUE TITLE
Support for downloading videos and offline playback

### ISSUE THREAD

**Issue by Cubxity** [flutter/flutter#59759]
## Use case
This would be useful for video streaming applications to allow users to be able to download and play videos offline. Downloading them externally would not provide as smooth experience as downloading directly with, for example, ExoPlayer.

## Proposal
I propose that `video_player` should have an API for managing downloads and the ability to play downloaded videos.

**Comment by Cubxity** [comment]
Hey, are there any updates regarding this?

**Comment by gkocyigit** [comment]
Is there any updates?

**Comment by seanjackson** [comment]
Good day mates is there any video player in flutter that can do this now,Currently facing the same issue

**Comment by a1rwulf** [comment]
As I haven't got any feedback on above, here is the implementation I have done meanwhile as a reference which is used in a real world application:
https://github.com/a1rwulf/flutter_packages/commits/feature_offline-playback/

Is there any interest in this feature for upstream video_player?

I decided to go for the very direct approach adding the download functionality to the video_player interface.
Ideally there would be a way to implement a plugin on top of video_player to add this functionality instead, so I'm open for feedback in case there is upstream interest.

**Comment by a1rwulf** [comment]
@Rexios80 you are correct, in case you want to play a normal movie file, it's possible using `VideoPlayerController.file`.
However OP is asking about a smoother experience for streaming applications, where I assume we talk for example about HLS streams and not normal video files.
That leads you to this problem: https://github.com/flutter/flutter/issues/101166

I worked out a solution as mentioned in my previous comment:
> As I haven't got any feedback on above, here is the implementation I have done meanwhile as a reference which is used in a real world application: https://github.com/a1rwulf/flutter_packages/commits/feature_offline-playback/
> 

But it's hacky and can't be upstreamed as is, so it would need a design document proposal with a clean separated interface.
As I'm not exactly sure how it can be done in a better way, I don't know when/if I have time to work on this.

Hope that helps.

**Comment by a1rwulf** [comment]
I finally came up with a way to support this using a separate package, which requires only a minimal change to video_player_android to inject a custom VideoAssetProvider.
This opens the way for a separate plugin that handles all download specifics while also Android would be able to play from the AssetProvider provided by said plugin.
iOS works without any changes, because the iOS player handles this differently.
